runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/hashing/HashFunction.kt

```diff
@@ -70,3 +71,60 @@ public fun String.toHashFunction(): HashFunction? = when (this.lowercase()) {
     "md5" -> Md5()
     else -> null
 }
+
+/**
+ * @return The [HashFunction] which is represented by this string, or an exception if none match.
+ */
+@InternalApi
+public fun String.toHashFunctionOrThrow(): HashFunction =
+    toHashFunction() ?: throw ClientException("Checksum algorithm is not supported: $this")
+
+/**
+ * @return If the [HashFunction] is supported by flexible checksums
+ */
+@InternalApi
+public val HashFunction.isSupportedForFlexibleChecksums: Boolean get() =
+    algorithmsSupportedForFlexibleChecksums.contains(this::class.simpleName)
```

Comment:
correctness: Relying on `this::class.simpleName` seems unsafe. 

Better solution:
```
@InternalApi
public val HashFunction.isSupportedForFlexibleChecksums: Boolean 
    get() = when (this) {
        is Crc32, is Crc32c, is Sha1, is Sha256 -> true
        else -> false
    }
```

Reply:
Yeah agree, that's what we had before but we would have to keep track of what checksum algorithms are supported for flexible checksums in two places. In `HashFunction.isSupportedForFlexibleChecksums` & `algorithmsSupportedForFlexibleChecksums`. I made a tradeoff here, it seems more risky to possibly miss updating what algorithms we support in both locations than relying on the class name. Plus we have unit tests that should lessen some of our concerns here

Reply:
The only other use of `algorithmsSupportedForFlexibleChecksums` is in a log message. I'd rather remove the log message than keep this class reflection